### PR TITLE
migrations: Update `dirty database` docs

### PR DESCRIPTION
**Pull request by efritz**
This PR is a rough docs update on how to handle a dirty database. This PR adds a new commands `sg migration add-log`, and `migrator add-log`, which can be used by developers or site-admins to help resolve the dirty database conditions with the new migration infrastructure.

**Reviewers**: Please give special attention to the doc page and suggest where I can steal copy from other migrations work we've been doing lately.

**Edit**: Non-docs work was pulled out into https://github.com/sourcegraph/sourcegraph/pull/30863.

### CODE REVIEW

File: doc/admin/install/kubernetes/update.md [hunk 1]
```diff
@@ -1,3 +1,5 @@
+TODO: update file
```
Comment (efritz) [inline doc/admin/install/kubernetes/update.md:1]: - [ ] Add a "version" command to the migrator

File: doc/admin/how-to/dirty_database.md [hunk 2]
```diff
@@ -16,81 +21,91 @@ Resolving this error requires discovering which migration file failed to run, an
...
+2. **Run the code from the identified `<version>/up.sql` file explicitly using the `psql` CLI:**
    * It’s possible that one or more commands from the migration ran successfully already. In these cases you may need to run the sql transaction in pieces. For example if a migration file creates multiple indexes and one index already exists you'll need to manually run this transaction skipping that line or adding `IF NOT EXISTS` to the transaction.
    * If you’re running into unique index creation errors because of duplicate values please let us know at support@sourcegraph.com or via your enterprise support channel.
    * There may be other error cases that don't have an easy admin-only resolution, in these cases please let us know at support@sourcegraph.com or via your enterprise support channel.
 
-### 3. Verify database is clean and declare `dirty=false`
+### 3. Add a migration log entry
+
+1. **Ensure the migration applied, then signal that the migration has been run.**
+   * Run the `migrator` instance against your database to create an explicit migration log.
```
Comment (efritz) [inline doc/admin/how-to/dirty_database.md:71]: Looking for specific suggestions on how to organize this. Also see [this related PR comment](https://github.com/sourcegraph/sourcegraph/pull/30776#discussion_r801218335).

**Review by mrnugget — APPROVED** [review]

**Comment by kevinwojo** [comment]
The `sg` commands should reduce the burden on our on-prem site admins. Really like this direction.

File: doc/admin/how-to/dirty_database.md [hunk 3]
```diff
@@ -4,10 +4,15 @@ This document will take you through how to resolve a 'dirty database' error. Dur
 
 The error will look something like this:
 
-```log
```
[the remark this answers is not in the document]
Comment (DaedalusG) [reply doc/admin/how-to/dirty_database.md:7]: ^ 2nd and my preference would be a separate page for the older versions

File: doc/admin/how-to/dirty_database.md [hunk 4]
```diff
@@ -16,81 +21,91 @@ Resolving this error requires discovering which migration file failed to run, an
...
+ORDER BY version
 ```
-This indicates that migration `1528395539` was running, but has not yet completed. 
-
-_Note: for codeintel the schema version table is called `codeintel_schema_migrations` and for codeinsights its called `codeinsights_schema_migrations`_
 
 ### 2. Run the sql queries to finish incomplete migrations
 
-Sourcegraph's migration files take for form of `sql` files following the snake case naming schema `<version>_<description>.<up or down>.sql` and can be found [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/migrations) in subdirectories for the specific database. _Note frontend is the pgsql database_.
+Migration definitions for each database schema can be found in the children of the [`migrations/` directory](https://github.com/sourcegraph/sourcegraph/tree/main/migrations).
```
Comment (DaedalusG) [inline doc/admin/how-to/dirty_database.md:59]: An example of what this looks like would be really nice

File: doc/admin/how-to/dirty_database.md [hunk 5]
```diff
@@ -16,81 +21,91 @@ Resolving this error requires discovering which migration file failed to run, an
...
    * There may be other error cases that don't have an easy admin-only resolution, in these cases please let us know at support@sourcegraph.com or via your enterprise support channel.
 
-### 3. Verify database is clean and declare `dirty=false`
+### 3. Add a migration log entry
+
+1. **Ensure the migration applied, then signal that the migration has been run.**
+   * Run the `migrator` instance against your database to create an explicit migration log.
+
+    ```bash
+    export SOURCEGRAPH_VERSION="The version you are upgrading to"
```
[the remark this answers is not in the document]
Comment (DaedalusG) [reply doc/admin/how-to/dirty_database.md:74]: An example version "version ex: v3.38.1" might be nice just to be extra explicit -- I like to use <> or $ type notations for when something required user specification in a code block

[the remark this answers is not in the document]
Comment (DaedalusG) [reply doc/admin/how-to/dirty_database.md:74]: ^^ also nitpicky

**Review by DaedalusG — APPROVED** [review]
This looks really great -- I know its the right thing -- but its sad seeing all that complicated procedure documentation be put to rest 😢 🤣

**Comment by mrnugget** [comment]
> The `sg` commands should reduce the burden on our on-prem site admins. Really like this direction.

Wait, where do site admins use `sg`? Or did you mean the `migrator`?

Because so far `sg` has been an internal-only tool to do Sourcegraph development. If we start pointing site-admins at it, we need to change how we handle releases/testing/QA etc. in `sg`.

**Comment by mrnugget** [comment]
Yup, that's exactly what I thought. That's why I was a bit surprised by @kevinwojo's comment. I don't see `sg` in this PR nor did I see you advertise it anywhere else as something that site-admins use.

So I guess it was a misunderstanding and `migrator` was meant.

File: doc/admin/how-to/dirty_database.md [hunk 6]
```diff
@@ -16,81 +21,91 @@ Resolving this error requires discovering which migration file failed to run, an
...
-   * _Note: migrations in this directory are specific to the `pgsql` frontend database, learn about other databases in the [detailed steps to resolution](#detailed-steps-to-resolve)_
-3. **Run the code there explicitly.**
-4. **Manually clear the dirty flag on the `schema_migrations` table.**
-5. **Start up again and the remaining migrations should succeed, otherwise repeat.**
+When migrations run, the `migration_logs` table is updated. Before each migration attempt, a new row is inserted indicating the migration version and direction and the start time. Once the migration is complete (or fails), the row is updated with the finished time and message with details about any error that occurred.
 
-## Detailed Steps to resolve
+A failed migration may have explicitly failed due to a SQL/environment error, or it may fail if the migrator instance dies before updating its log entry. In this case, the validation mechanism that runs on app startup will wait for running migrators to complete their current work. Pending _but inactive_ migration attempts will be correct reported as failed, as seen again below.
```
[the remark this answers is not in the document]
Comment (efritz) [reply doc/admin/how-to/dirty_database.md:38]: Tried to rework this a bit.

File: doc/admin/how-to/dirty_database.md (hunk 4 above)
Comment (efritz) [reply doc/admin/how-to/dirty_database.md:67]: Marked a temporary todo while I rework this.

**Review by mrnugget — APPROVED** [review]

File: doc/admin/how-to/dirty_database.md [hunk 7]
```diff
@@ -1,97 +1,106 @@
...
-FROM pg_class pc
-JOIN pg_index pi ON pi.indexrelid = pc.oid
-WHERE
-    NOT indisvalid AND
-    NOT EXISTS (SELECT 1 FROM pg_stat_progress_create_index pci WHERE pci.index_relid = pi.indexrelid)
-```
-Additionally Grafana will alert you of an index is in this state. _The Grafana alert can be found under it's database's charts._ Ex: `Site Admin > Monitoring > Postgres > Invalid Indexes (unusable by query planner)`
+### 3. Add a migration log entry
+
+**Ensure the migration applied, then signal that the migration has been run**. Run the `migrator` instance against your database to create an explicit migration log. For the following, consult the [Kubernetes](./manual_database_migrations.md#kubernetes) or [Docker-compose](./manual_database_migrations.md#docker-compose) instructions on manually run database operations. The specific migrator command to run is `add-log -db=<schema> -version=<version>`. This indiciates to future migrator and application instances that a complete application of the migration at that version has been run.
```
Comment (chrismwendt) [inline doc/admin/how-to/dirty_database.md:102]:
```suggestion
**Ensure the migration applied, then signal that the migration has been run**. Run the `migrator` instance against your database to create an explicit migration log. For the following, consult the [Kubernetes](./manual_database_migrations.md#kubernetes) or [Docker-compose](./manual_database_migrations.md#docker-compose) or [local development](./manual_database_migrations.md#local-development) instructions on manually run database operations. The specific migrator command to run is `add-log -db=<schema> -version=<version>`. This indiciates to future migrator and application instances that a complete application of the migration at that version has been run.
```

**Review by DaedalusG — APPROVED** [review]
Looks good -- everyone seems to have ironed out the grammar too 😅

File: doc/admin/how-to/dirty_database.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (efritz) [reply doc/admin/how-to/dirty_database.md]: Updated in UI (couldn't apply as diff is outdated).

**Pull request merged**